helper/schema/resource.go

```diff
@@ -69,6 +96,13 @@ type DeleteFunc func(*ResourceData, interface{}) error
 // See Resource documentation.
 type ExistsFunc func(*ResourceData, interface{}) (bool, error)
 
+// See Resource documentation for more details.
+type StateMigrationContext struct {
+	StateSchemaVersion int
+	State              *terraform.InstanceState
+}
+type StateMigrateFunc func(*StateMigrationContext, interface{}) error
```

Comment:
I think this should return `(*InstanceState, error)`. They might just modify the context state but I think that makes the API clearer. 

Second question: why the context struct? What future thing are you thinking of here?